src/azul/chalice.py

```diff
@@ -191,10 +191,12 @@ def base_url(self) -> furl:
         """
         The base URL of this application. Callers can safely modify the returned
         `furl` instance.
+        The base URL has no path and should be set by caller.
         """
         if self.current_request is None:
             # Invocation via AWS StepFunctions
             self_url = furl(config.service_endpoint())
+            assert self_url.path == 'localhost'
```

Comment:
This can't be right. `localhost` is a hostname, not a path.

Reply:
This was a case observed mainly through the `make openapi` step in gh actions. I printed both the path and the url contents and they both were `localhost`. I can leave out this assertion if it is unnecessary as the assertion to check for no path is isolated in the else clause of the if statements. If not, should I go with `assert str(self_url) == 'localhost'` instead?

Reply:
This is because we specify the patch as `localhost` without a scheme. Try adding this patch as a separate commit to your PR, place it before the commit that resolves #3398.
 

```
Index: scripts/generate_openapi_document.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/generate_openapi_document.py b/scripts/generate_openapi_document.py
--- a/scripts/generate_openapi_document.py	(revision fe9496d1509e1257b70be5926b166159ac93a66f)
+++ b/scripts/generate_openapi_document.py	(date 1649105068402)
@@ -39,8 +39,8 @@
             assert config.service_name == 'azul_service'
             with patch.object(target=config,
                               attribute='service_endpoint',
-                              return_value='localhost'):
-                assert config.service_endpoint() == 'localhost'
+                              return_value='http://localhost'):
+                assert config.service_endpoint() == 'http://localhost'
                 app_module = load_app_module('service')
                 app_spec = app_module.app.spec()
                 doc_path = os.path.join(config.project_root, 'lambdas/service/openapi.json')
Index: lambdas/service/openapi.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/lambdas/service/openapi.json b/lambdas/service/openapi.json
--- a/lambdas/service/openapi.json	(revision fe9496d1509e1257b70be5926b166159ac93a66f)
+++ b/lambdas/service/openapi.json	(date 1649105563709)
@@ -39967,7 +39967,7 @@
     },
     "servers": [
         {
-            "url": "/"
+            "url": "http://localhost/"
         }
     ],
     "components": {

```

Then you can move the assertion that you added

```
Index: src/azul/chalice.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/chalice.py b/src/azul/chalice.py
--- a/src/azul/chalice.py	(revision fe9496d1509e1257b70be5926b166159ac93a66f)
+++ b/src/azul/chalice.py	(date 1649105551826)
@@ -209,9 +209,9 @@
                 # Invocation via API Gateway
                 pass
             self_url = furl(scheme=scheme, netloc=self.current_request.headers['host'])
-            assert self_url.path == ''
         else:
             assert False, self.current_request
+        assert self_url.path == ''
         return self_url
 
     def _register_spec(self,
```